{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction to Python for Data Analsyis\n",
    "\n",
    "### Module 1: DataFrames and Python\n",
    "Duration: ~ 45'\n",
    "\n",
    "In this module, we will explore:\n",
    "\n",
    "- What is a pandas Dataframe\n",
    "- How to read and export data in pandas\n",
    "- How to perform basic operations with a pandas DataFrame, such as:\n",
    "    - Sorting\n",
    "    - Locating by index and position\n",
    "- Perform more complex operatios like grouping and filtering\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What is pandas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pandas is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool,\n",
    "built on top of the Python programming language. In particular, it offers data structures and operations for manipulating numerical tables and time series. \n",
    "\n",
    "Its core object is called *DataFrame*; a data structure that organizes data into a 2-dimensional table of rows and columns, much like a spreadsheet. Its advantage over traditional arrays (such as ones used by *numpy*) is that each column can contain information **of different type**, e.g. one columsn can containg strings (names, occupations, martial status) and adjacrnt columns can be, among others, intgeres, floats, datetimes etc.\n",
    "\n",
    "It is built **on top** of numpy (the core numerical library of python) which makes it highly efficient Also, some core parts of pandas are written in Cython (a superset of the Python language that additionally supports calling C functions and declaring C types on variables and class attributes)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading pandas\n",
    "\n",
    "Once installed, you can import it e.g. using the alias `pd` as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading datasets with `pandas`\n",
    "\n",
    "We are going to use the METABRIC dataset, on open source dataset which contains targeted sequencing, clinical and genomic data from patients with breast cancer. More information about the dataset can be found [here](https://ega-archive.org/studies/EGAS00000000083) and [here](https://www.kaggle.com/datasets/raghadalharbi/breast-cancer-gene-expression-profiles-metabric).\n",
    "\n",
    "Pandas allows importing data from various file formats such as csv, xls, json, sql. They all follow the same pattern: `.read_{format_of_the_file}`, e.g. `.read_sql()`, `.read_xls()` etc.\n",
    "\n",
    "To read a csv file, use the method `.read_csv()`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "metabric = pd.read_csv(\"../data/metabric_clinical_and_expression_data.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you forget to include `../data/` above, or if you include it but your copy of the file is saved somewhere else, you will get an error that ends with a line like this: \n",
    "\n",
    "`FileNotFoundError: File b'metabric_clinical_and_expression_data.csv' does not exist`\n",
    "\n",
    "Generally, rows in a `DataFrame` are the **observations** (patients in the case of METABRIC) whereas columns are known as the observed **variables** (Cohort, Age_at_diagnosis ...). \n",
    "\n",
    "Looking at the column on the far left, you can see the row names of the DataFrame `metabric` assigned using the known 0-based indexing used in Python.\n",
    "\n",
    "Note that the `.read_csv()` method is not limited to reading csv files. For example, you can also read Tab Separated Value (TSV) files by adding the argument `sep='\\t'`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exploring the dataset\n",
    "\n",
    "DataFrames in pandas are classes, which means they carry specific attributes. \n",
    "\n",
    "**Methods** are functions that are associated with a DataFrame. Because they are functions, you do use () to call them, and can have arguments added inside the parentheses to control their behaviour. For example, the `.info()` command we executed previously was a method.\n",
    "\n",
    "\n",
    "The most important methods ones to initally examine a dataset are:\n",
    "\n",
    "- `.info()`: We can view basic information about our DataFrame object, like the type of each column and the number of missing entries.\n",
    "- `.describe()`: We can quickly view some statistical meassures for our numerical variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 1904 entries, 0 to 1903\n",
      "Data columns (total 32 columns):\n",
      " #   Column                         Non-Null Count  Dtype  \n",
      "---  ------                         --------------  -----  \n",
      " 0   Patient_ID                     1904 non-null   object \n",
      " 1   Cohort                         1904 non-null   int64  \n",
      " 2   Age_at_diagnosis               1904 non-null   float64\n",
      " 3   Survival_time                  1904 non-null   float64\n",
      " 4   Survival_status                1904 non-null   object \n",
      " 5   Vital_status                   1903 non-null   object \n",
      " 6   Chemotherapy                   1904 non-null   object \n",
      " 7   Radiotherapy                   1904 non-null   object \n",
      " 8   Tumour_size                    1884 non-null   float64\n",
      " 9   Tumour_stage                   1403 non-null   float64\n",
      " 10  Neoplasm_histologic_grade      1832 non-null   float64\n",
      " 11  Lymph_nodes_examined_positive  1904 non-null   int64  \n",
      " 12  Lymph_node_status              1904 non-null   int64  \n",
      " 13  Cancer_type                    1889 non-null   object \n",
      " 14  ER_status                      1904 non-null   object \n",
      " 15  PR_status                      1904 non-null   object \n",
      " 16  HER2_status                    1904 non-null   object \n",
      " 17  HER2_status_measured_by_SNP6   1904 non-null   object \n",
      " 18  PAM50                          1904 non-null   object \n",
      " 19  3-gene_classifier              1700 non-null   object \n",
      " 20  Nottingham_prognostic_index    1904 non-null   float64\n",
      " 21  Cellularity                    1850 non-null   object \n",
      " 22  Integrative_cluster            1904 non-null   object \n",
      " 23  Mutation_count                 1859 non-null   float64\n",
      " 24  ESR1                           1904 non-null   float64\n",
      " 25  ERBB2                          1904 non-null   float64\n",
      " 26  PGR                            1904 non-null   float64\n",
      " 27  TP53                           1904 non-null   float64\n",
      " 28  PIK3CA                         1904 non-null   float64\n",
      " 29  GATA3                          1904 non-null   float64\n",
      " 30  FOXA1                          1904 non-null   float64\n",
      " 31  MLPH                           1904 non-null   float64\n",
      "dtypes: float64(15), int64(3), object(14)\n",
      "memory usage: 476.1+ KB\n"
     ]
    }
   ],
   "source": [
    "metabric.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cohort</th>\n",
       "      <th>Age_at_diagnosis</th>\n",
       "      <th>Survival_time</th>\n",
       "      <th>Tumour_size</th>\n",
       "      <th>Tumour_stage</th>\n",
       "      <th>Neoplasm_histologic_grade</th>\n",
       "      <th>Lymph_nodes_examined_positive</th>\n",
       "      <th>Lymph_node_status</th>\n",
       "      <th>Nottingham_prognostic_index</th>\n",
       "      <th>Mutation_count</th>\n",
       "      <th>ESR1</th>\n",
       "      <th>ERBB2</th>\n",
       "      <th>PGR</th>\n",
       "      <th>TP53</th>\n",
       "      <th>PIK3CA</th>\n",
       "      <th>GATA3</th>\n",
       "      <th>FOXA1</th>\n",
       "      <th>MLPH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>1904.000000</td>\n",
       "      <td>1904.000000</td>\n",
       "      <td>1904.000000</td>\n",
       "      <td>1884.000000</td>\n",
       "      <td>1403.000000</td>\n",
       "      <td>1832.000000</td>\n",
       "      <td>1904.000000</td>\n",
       "      <td>1904.000000</td>\n",
       "      <td>1904.000000</td>\n",
       "      <td>1859.000000</td>\n",
       "      <td>1904.000000</td>\n",
       "      <td>1904.000000</td>\n",
       "      <td>1904.000000</td>\n",
       "      <td>1904.000000</td>\n",
       "      <td>1904.000000</td>\n",
       "      <td>1904.000000</td>\n",
       "      <td>1904.000000</td>\n",
       "      <td>1904.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>2.643908</td>\n",
       "      <td>61.087054</td>\n",
       "      <td>125.121324</td>\n",
       "      <td>26.238726</td>\n",
       "      <td>1.750535</td>\n",
       "      <td>2.415939</td>\n",
       "      <td>2.002101</td>\n",
       "      <td>1.639706</td>\n",
       "      <td>4.033019</td>\n",
       "      <td>5.697687</td>\n",
       "      <td>9.607824</td>\n",
       "      <td>10.765364</td>\n",
       "      <td>6.237203</td>\n",
       "      <td>6.197967</td>\n",
       "      <td>5.970097</td>\n",
       "      <td>9.502910</td>\n",
       "      <td>10.800526</td>\n",
       "      <td>11.362384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>1.228615</td>\n",
       "      <td>12.978711</td>\n",
       "      <td>76.334148</td>\n",
       "      <td>15.160976</td>\n",
       "      <td>0.628999</td>\n",
       "      <td>0.650612</td>\n",
       "      <td>4.079993</td>\n",
       "      <td>0.743809</td>\n",
       "      <td>1.144492</td>\n",
       "      <td>4.058778</td>\n",
       "      <td>2.133827</td>\n",
       "      <td>1.357359</td>\n",
       "      <td>1.020871</td>\n",
       "      <td>0.401864</td>\n",
       "      <td>0.352549</td>\n",
       "      <td>1.502636</td>\n",
       "      <td>1.754282</td>\n",
       "      <td>1.687555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>21.930000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>5.217238</td>\n",
       "      <td>6.372949</td>\n",
       "      <td>4.860645</td>\n",
       "      <td>5.201128</td>\n",
       "      <td>5.158697</td>\n",
       "      <td>5.277722</td>\n",
       "      <td>5.184945</td>\n",
       "      <td>5.323652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>51.375000</td>\n",
       "      <td>60.825000</td>\n",
       "      <td>17.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.046000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>8.092992</td>\n",
       "      <td>9.969681</td>\n",
       "      <td>5.408728</td>\n",
       "      <td>5.930335</td>\n",
       "      <td>5.735007</td>\n",
       "      <td>8.767954</td>\n",
       "      <td>10.829777</td>\n",
       "      <td>11.042871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>3.000000</td>\n",
       "      <td>61.770000</td>\n",
       "      <td>115.616667</td>\n",
       "      <td>23.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>4.042000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>10.252166</td>\n",
       "      <td>10.530301</td>\n",
       "      <td>5.877591</td>\n",
       "      <td>6.185873</td>\n",
       "      <td>5.938094</td>\n",
       "      <td>9.911805</td>\n",
       "      <td>11.367947</td>\n",
       "      <td>11.873967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>3.000000</td>\n",
       "      <td>70.592500</td>\n",
       "      <td>184.716667</td>\n",
       "      <td>30.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>5.040250</td>\n",
       "      <td>7.000000</td>\n",
       "      <td>11.268331</td>\n",
       "      <td>11.159306</td>\n",
       "      <td>6.899220</td>\n",
       "      <td>6.456987</td>\n",
       "      <td>6.148720</td>\n",
       "      <td>10.560220</td>\n",
       "      <td>11.779545</td>\n",
       "      <td>12.396317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>5.000000</td>\n",
       "      <td>96.290000</td>\n",
       "      <td>355.200000</td>\n",
       "      <td>182.000000</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>45.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>6.360000</td>\n",
       "      <td>80.000000</td>\n",
       "      <td>13.265184</td>\n",
       "      <td>14.643900</td>\n",
       "      <td>9.932115</td>\n",
       "      <td>7.921411</td>\n",
       "      <td>8.708396</td>\n",
       "      <td>12.812082</td>\n",
       "      <td>13.127682</td>\n",
       "      <td>14.432001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Cohort  Age_at_diagnosis  Survival_time  Tumour_size  \\\n",
       "count  1904.000000       1904.000000    1904.000000  1884.000000   \n",
       "mean      2.643908         61.087054     125.121324    26.238726   \n",
       "std       1.228615         12.978711      76.334148    15.160976   \n",
       "min       1.000000         21.930000       0.000000     1.000000   \n",
       "25%       1.000000         51.375000      60.825000    17.000000   \n",
       "50%       3.000000         61.770000     115.616667    23.000000   \n",
       "75%       3.000000         70.592500     184.716667    30.000000   \n",
       "max       5.000000         96.290000     355.200000   182.000000   \n",
       "\n",
       "       Tumour_stage  Neoplasm_histologic_grade  Lymph_nodes_examined_positive  \\\n",
       "count   1403.000000                1832.000000                    1904.000000   \n",
       "mean       1.750535                   2.415939                       2.002101   \n",
       "std        0.628999                   0.650612                       4.079993   \n",
       "min        0.000000                   1.000000                       0.000000   \n",
       "25%        1.000000                   2.000000                       0.000000   \n",
       "50%        2.000000                   3.000000                       0.000000   \n",
       "75%        2.000000                   3.000000                       2.000000   \n",
       "max        4.000000                   3.000000                      45.000000   \n",
       "\n",
       "       Lymph_node_status  Nottingham_prognostic_index  Mutation_count  \\\n",
       "count        1904.000000                  1904.000000     1859.000000   \n",
       "mean            1.639706                     4.033019        5.697687   \n",
       "std             0.743809                     1.144492        4.058778   \n",
       "min             1.000000                     1.000000        1.000000   \n",
       "25%             1.000000                     3.046000        3.000000   \n",
       "50%             1.000000                     4.042000        5.000000   \n",
       "75%             2.000000                     5.040250        7.000000   \n",
       "max             3.000000                     6.360000       80.000000   \n",
       "\n",
       "              ESR1        ERBB2          PGR         TP53       PIK3CA  \\\n",
       "count  1904.000000  1904.000000  1904.000000  1904.000000  1904.000000   \n",
       "mean      9.607824    10.765364     6.237203     6.197967     5.970097   \n",
       "std       2.133827     1.357359     1.020871     0.401864     0.352549   \n",
       "min       5.217238     6.372949     4.860645     5.201128     5.158697   \n",
       "25%       8.092992     9.969681     5.408728     5.930335     5.735007   \n",
       "50%      10.252166    10.530301     5.877591     6.185873     5.938094   \n",
       "75%      11.268331    11.159306     6.899220     6.456987     6.148720   \n",
       "max      13.265184    14.643900     9.932115     7.921411     8.708396   \n",
       "\n",
       "             GATA3        FOXA1         MLPH  \n",
       "count  1904.000000  1904.000000  1904.000000  \n",
       "mean      9.502910    10.800526    11.362384  \n",
       "std       1.502636     1.754282     1.687555  \n",
       "min       5.277722     5.184945     5.323652  \n",
       "25%       8.767954    10.829777    11.042871  \n",
       "50%       9.911805    11.367947    11.873967  \n",
       "75%      10.560220    11.779545    12.396317  \n",
       "max      12.812082    13.127682    14.432001  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabric.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can access the dimensions of your DataFrame using the `.shape` attribute. The first value is the number of rows, and the second the number of columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1904, 32)\n",
      "Number of rows: 1904\n",
      "Number of columns: 32\n"
     ]
    }
   ],
   "source": [
    "print(metabric.shape)\n",
    "print(f\"Number of rows: {metabric.shape[0]}\")\n",
    "print(f\"Number of columns: {metabric.shape[1]}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The row and column names can be accessed using the attributes `.index` and `.columns` respectively:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RangeIndex(start=0, stop=1904, step=1)\n",
      "Index(['Patient_ID', 'Cohort', 'Age_at_diagnosis', 'Survival_time',\n",
      "       'Survival_status', 'Vital_status', 'Chemotherapy', 'Radiotherapy',\n",
      "       'Tumour_size', 'Tumour_stage', 'Neoplasm_histologic_grade',\n",
      "       'Lymph_nodes_examined_positive', 'Lymph_node_status', 'Cancer_type',\n",
      "       'ER_status', 'PR_status', 'HER2_status', 'HER2_status_measured_by_SNP6',\n",
      "       'PAM50', '3-gene_classifier', 'Nottingham_prognostic_index',\n",
      "       'Cellularity', 'Integrative_cluster', 'Mutation_count', 'ESR1', 'ERBB2',\n",
      "       'PGR', 'TP53', 'PIK3CA', 'GATA3', 'FOXA1', 'MLPH'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(metabric.index)\n",
    "print(metabric.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Accesing individual columns of a pandas dataframe can be achieved in two ways:\n",
    "\n",
    "- The first is using the name of the DataFrame `metabric` followed by a `.` and then followed by the name of the column. \n",
    "- The second is using square brackets.\n",
    "\n",
    "*Note: If your column name contains spaces, the first method is **not** applicable*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0       140.500000\n",
      "1        84.633333\n",
      "2       163.700000\n",
      "3       164.933333\n",
      "4        41.366667\n",
      "           ...    \n",
      "1899    196.866667\n",
      "1900     44.733333\n",
      "1901    175.966667\n",
      "1902     86.233333\n",
      "1903    201.900000\n",
      "Name: Survival_time, Length: 1904, dtype: float64 0       140.500000\n",
      "1        84.633333\n",
      "2       163.700000\n",
      "3       164.933333\n",
      "4        41.366667\n",
      "           ...    \n",
      "1899    196.866667\n",
      "1900     44.733333\n",
      "1901    175.966667\n",
      "1902     86.233333\n",
      "1903    201.900000\n",
      "Name: Survival_time, Length: 1904, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "print(metabric.Survival_time, metabric[\"Survival_time\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also compute metrics on specific columns or on the entire DataFrame:\n",
    "\n",
    "*Note: Using metrics methods on whole DataFrames that contain categorical columns will ommit them aytomatically. Depending on the pandas version, this will result to an warning.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "125.12132352924738\n",
      "76.33414828620853\n",
      "Cohort                             2.643908\n",
      "Age_at_diagnosis                  61.087054\n",
      "Survival_time                    125.121324\n",
      "Tumour_size                       26.238726\n",
      "Tumour_stage                       1.750535\n",
      "Neoplasm_histologic_grade          2.415939\n",
      "Lymph_nodes_examined_positive      2.002101\n",
      "Lymph_node_status                  1.639706\n",
      "Nottingham_prognostic_index        4.033019\n",
      "Mutation_count                     5.697687\n",
      "ESR1                               9.607824\n",
      "ERBB2                             10.765364\n",
      "PGR                                6.237203\n",
      "TP53                               6.197967\n",
      "PIK3CA                             5.970097\n",
      "GATA3                              9.502910\n",
      "FOXA1                             10.800526\n",
      "MLPH                              11.362384\n",
      "dtype: float64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/tq/8nt7c8ld587csq8jpk2dycbr0000gn/T/ipykernel_1366/2193329495.py:3: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.\n",
      "  print(metabric.mean())\n"
     ]
    }
   ],
   "source": [
    "print(metabric['Survival_time'].mean())\n",
    "print(metabric['Survival_time'].std())\n",
    "print(metabric.mean())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Selecting columns and rows\n",
    "\n",
    "The [pandas cheat sheet](https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf) can be very helpful for recalling basic pandas operations.\n",
    "\n",
    "To select rows and columns in a DataFrame, we use square brackets `[ ]`. There are two ways to do this: with **positional** indexing, which uses index numbers, and **label-based** indexing which uses column or row names.\n",
    "\n",
    "To select the first three rows using their numeric index (The colon `:` defines the range \"from the beggining until\"):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Cohort</th>\n",
       "      <th>Age_at_diagnosis</th>\n",
       "      <th>Survival_time</th>\n",
       "      <th>Survival_status</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>Chemotherapy</th>\n",
       "      <th>Radiotherapy</th>\n",
       "      <th>Tumour_size</th>\n",
       "      <th>Tumour_stage</th>\n",
       "      <th>...</th>\n",
       "      <th>Integrative_cluster</th>\n",
       "      <th>Mutation_count</th>\n",
       "      <th>ESR1</th>\n",
       "      <th>ERBB2</th>\n",
       "      <th>PGR</th>\n",
       "      <th>TP53</th>\n",
       "      <th>PIK3CA</th>\n",
       "      <th>GATA3</th>\n",
       "      <th>FOXA1</th>\n",
       "      <th>MLPH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MB-0000</td>\n",
       "      <td>1</td>\n",
       "      <td>75.65</td>\n",
       "      <td>140.500000</td>\n",
       "      <td>LIVING</td>\n",
       "      <td>Living</td>\n",
       "      <td>NO</td>\n",
       "      <td>YES</td>\n",
       "      <td>22.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4ER+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.929817</td>\n",
       "      <td>9.333972</td>\n",
       "      <td>5.680501</td>\n",
       "      <td>6.338739</td>\n",
       "      <td>5.704157</td>\n",
       "      <td>6.932146</td>\n",
       "      <td>7.953794</td>\n",
       "      <td>9.729728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MB-0002</td>\n",
       "      <td>1</td>\n",
       "      <td>43.19</td>\n",
       "      <td>84.633333</td>\n",
       "      <td>LIVING</td>\n",
       "      <td>Living</td>\n",
       "      <td>NO</td>\n",
       "      <td>YES</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4ER+</td>\n",
       "      <td>2.0</td>\n",
       "      <td>10.047059</td>\n",
       "      <td>9.729606</td>\n",
       "      <td>7.505424</td>\n",
       "      <td>6.192507</td>\n",
       "      <td>5.757727</td>\n",
       "      <td>11.251197</td>\n",
       "      <td>11.843989</td>\n",
       "      <td>12.536570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MB-0005</td>\n",
       "      <td>1</td>\n",
       "      <td>48.87</td>\n",
       "      <td>163.700000</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>YES</td>\n",
       "      <td>NO</td>\n",
       "      <td>15.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>2.0</td>\n",
       "      <td>10.041281</td>\n",
       "      <td>9.725825</td>\n",
       "      <td>7.376123</td>\n",
       "      <td>6.404516</td>\n",
       "      <td>6.751566</td>\n",
       "      <td>9.289758</td>\n",
       "      <td>11.698169</td>\n",
       "      <td>10.306115</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Patient_ID  Cohort  Age_at_diagnosis  Survival_time Survival_status  \\\n",
       "0    MB-0000       1             75.65     140.500000          LIVING   \n",
       "1    MB-0002       1             43.19      84.633333          LIVING   \n",
       "2    MB-0005       1             48.87     163.700000        DECEASED   \n",
       "\n",
       "      Vital_status Chemotherapy Radiotherapy  Tumour_size  Tumour_stage  ...  \\\n",
       "0           Living           NO          YES         22.0           2.0  ...   \n",
       "1           Living           NO          YES         10.0           1.0  ...   \n",
       "2  Died of Disease          YES           NO         15.0           2.0  ...   \n",
       "\n",
       "   Integrative_cluster  Mutation_count       ESR1     ERBB2       PGR  \\\n",
       "0                 4ER+             NaN   8.929817  9.333972  5.680501   \n",
       "1                 4ER+             2.0  10.047059  9.729606  7.505424   \n",
       "2                    3             2.0  10.041281  9.725825  7.376123   \n",
       "\n",
       "       TP53    PIK3CA      GATA3      FOXA1       MLPH  \n",
       "0  6.338739  5.704157   6.932146   7.953794   9.729728  \n",
       "1  6.192507  5.757727  11.251197  11.843989  12.536570  \n",
       "2  6.404516  6.751566   9.289758  11.698169  10.306115  \n",
       "\n",
       "[3 rows x 32 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabric[:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And we can combine the row and columns selection like that:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    NaN\n",
       "1    2.0\n",
       "2    2.0\n",
       "Name: Mutation_count, dtype: float64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabric[:3]['Mutation_count']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "However the following does not work, as it would in languages like R:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "ename": "InvalidIndexError",
     "evalue": "(slice(None, 3, None), 'Mutation_count')",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "File \u001b[0;32m~/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/indexes/base.py:3621\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   <a href='file:///Users/leosouliotis/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3619'>3620</a>\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> <a href='file:///Users/leosouliotis/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3620'>3621</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_engine\u001b[39m.\u001b[39;49mget_loc(casted_key)\n\u001b[1;32m   <a href='file:///Users/leosouliotis/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3621'>3622</a>\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyError\u001b[39;00m \u001b[39mas\u001b[39;00m err:\n",
      "File \u001b[0;32mpandas/_libs/index.pyx:136\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/index.pyx:142\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: '(slice(None, 3, None), 'Mutation_count')' is an invalid key",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mInvalidIndexError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/Users/leosouliotis/Documents/ODSC/material/module_1.ipynb Cell 25'\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/leosouliotis/Documents/ODSC/material/module_1.ipynb#ch0000036?line=0'>1</a>\u001b[0m metabric[:\u001b[39m3\u001b[39;49m,\u001b[39m'\u001b[39;49m\u001b[39mMutation_count\u001b[39;49m\u001b[39m'\u001b[39;49m]\n",
      "File \u001b[0;32m~/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/frame.py:3505\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   <a href='file:///Users/leosouliotis/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/frame.py?line=3502'>3503</a>\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mcolumns\u001b[39m.\u001b[39mnlevels \u001b[39m>\u001b[39m \u001b[39m1\u001b[39m:\n\u001b[1;32m   <a href='file:///Users/leosouliotis/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/frame.py?line=3503'>3504</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> <a href='file:///Users/leosouliotis/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/frame.py?line=3504'>3505</a>\u001b[0m indexer \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mcolumns\u001b[39m.\u001b[39;49mget_loc(key)\n\u001b[1;32m   <a href='file:///Users/leosouliotis/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/frame.py?line=3505'>3506</a>\u001b[0m \u001b[39mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   <a href='file:///Users/leosouliotis/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/frame.py?line=3506'>3507</a>\u001b[0m     indexer \u001b[39m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m~/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/indexes/base.py:3628\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   <a href='file:///Users/leosouliotis/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3622'>3623</a>\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mKeyError\u001b[39;00m(key) \u001b[39mfrom\u001b[39;00m \u001b[39merr\u001b[39;00m\n\u001b[1;32m   <a href='file:///Users/leosouliotis/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3623'>3624</a>\u001b[0m     \u001b[39mexcept\u001b[39;00m \u001b[39mTypeError\u001b[39;00m:\n\u001b[1;32m   <a href='file:///Users/leosouliotis/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3624'>3625</a>\u001b[0m         \u001b[39m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   <a href='file:///Users/leosouliotis/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3625'>3626</a>\u001b[0m         \u001b[39m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   <a href='file:///Users/leosouliotis/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3626'>3627</a>\u001b[0m         \u001b[39m#  the TypeError.\u001b[39;00m\n\u001b[0;32m-> <a href='file:///Users/leosouliotis/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3627'>3628</a>\u001b[0m         \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_check_indexing_error(key)\n\u001b[1;32m   <a href='file:///Users/leosouliotis/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3628'>3629</a>\u001b[0m         \u001b[39mraise\u001b[39;00m\n\u001b[1;32m   <a href='file:///Users/leosouliotis/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3630'>3631</a>\u001b[0m \u001b[39m# GH#42269\u001b[39;00m\n",
      "File \u001b[0;32m~/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/indexes/base.py:5637\u001b[0m, in \u001b[0;36mIndex._check_indexing_error\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   <a href='file:///Users/leosouliotis/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=5632'>5633</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_check_indexing_error\u001b[39m(\u001b[39mself\u001b[39m, key):\n\u001b[1;32m   <a href='file:///Users/leosouliotis/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=5633'>5634</a>\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m is_scalar(key):\n\u001b[1;32m   <a href='file:///Users/leosouliotis/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=5634'>5635</a>\u001b[0m         \u001b[39m# if key is not a scalar, directly raise an error (the code below\u001b[39;00m\n\u001b[1;32m   <a href='file:///Users/leosouliotis/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=5635'>5636</a>\u001b[0m         \u001b[39m# would convert to numpy arrays and raise later any way) - GH29926\u001b[39;00m\n\u001b[0;32m-> <a href='file:///Users/leosouliotis/Documents/ODSC/.venv/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=5636'>5637</a>\u001b[0m         \u001b[39mraise\u001b[39;00m InvalidIndexError(key)\n",
      "\u001b[0;31mInvalidIndexError\u001b[0m: (slice(None, 3, None), 'Mutation_count')"
     ]
    }
   ],
   "source": [
    "metabric[:3,'Mutation_count']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To do **positional** indexing for both rows and columns, use `.iloc[]`. The first argument is the numeric index of the rows, and the second the numeric index of the columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    75.65\n",
       "1    43.19\n",
       "2    48.87\n",
       "Name: Age_at_diagnosis, dtype: float64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabric.iloc[:3,2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For **label-based** indexing, use `.loc[]` with the column and row names:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    75.65\n",
       "1    43.19\n",
       "2    48.87\n",
       "3    47.68\n",
       "Name: Age_at_diagnosis, dtype: float64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabric.loc[:3,\"Age_at_diagnosis\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note**: because the rows have numeric indices in this DataFrame, we may think that selecting rows with `.iloc[]` and `.loc[]` is same. As observed above, this is not the case.\n",
    "\n",
    "If you'd like to select more than one column:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cohort</th>\n",
       "      <th>Chemotherapy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>YES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>YES</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Cohort Chemotherapy\n",
       "0       1           NO\n",
       "1       1           NO\n",
       "2       1          YES\n",
       "3       1          YES"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabric.loc[:3, ['Cohort', 'Chemotherapy']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cohort</th>\n",
       "      <th>Age_at_diagnosis</th>\n",
       "      <th>Survival_time</th>\n",
       "      <th>Survival_status</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>Chemotherapy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>75.65</td>\n",
       "      <td>140.500000</td>\n",
       "      <td>LIVING</td>\n",
       "      <td>Living</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>43.19</td>\n",
       "      <td>84.633333</td>\n",
       "      <td>LIVING</td>\n",
       "      <td>Living</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>48.87</td>\n",
       "      <td>163.700000</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>YES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>47.68</td>\n",
       "      <td>164.933333</td>\n",
       "      <td>LIVING</td>\n",
       "      <td>Living</td>\n",
       "      <td>YES</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Cohort  Age_at_diagnosis  Survival_time Survival_status     Vital_status  \\\n",
       "0       1             75.65     140.500000          LIVING           Living   \n",
       "1       1             43.19      84.633333          LIVING           Living   \n",
       "2       1             48.87     163.700000        DECEASED  Died of Disease   \n",
       "3       1             47.68     164.933333          LIVING           Living   \n",
       "\n",
       "  Chemotherapy  \n",
       "0           NO  \n",
       "1           NO  \n",
       "2          YES  \n",
       "3          YES  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabric.loc[:3, 'Cohort':'Chemotherapy']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filtering rows\n",
    "\n",
    "You can choose rows from a DataFrame that match some specified criteria. The criteria are based on values of variables and can make use of comparison operators such as `==`, `>`, `<` and `!=`.\n",
    "\n",
    "For example, to filter `metabric` so that it only contains observations for those patients who died of breast cancer:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Cohort</th>\n",
       "      <th>Age_at_diagnosis</th>\n",
       "      <th>Survival_time</th>\n",
       "      <th>Survival_status</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>Chemotherapy</th>\n",
       "      <th>Radiotherapy</th>\n",
       "      <th>Tumour_size</th>\n",
       "      <th>Tumour_stage</th>\n",
       "      <th>...</th>\n",
       "      <th>Integrative_cluster</th>\n",
       "      <th>Mutation_count</th>\n",
       "      <th>ESR1</th>\n",
       "      <th>ERBB2</th>\n",
       "      <th>PGR</th>\n",
       "      <th>TP53</th>\n",
       "      <th>PIK3CA</th>\n",
       "      <th>GATA3</th>\n",
       "      <th>FOXA1</th>\n",
       "      <th>MLPH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MB-0005</td>\n",
       "      <td>1</td>\n",
       "      <td>48.87</td>\n",
       "      <td>163.700000</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>YES</td>\n",
       "      <td>NO</td>\n",
       "      <td>15.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>2.0</td>\n",
       "      <td>10.041281</td>\n",
       "      <td>9.725825</td>\n",
       "      <td>7.376123</td>\n",
       "      <td>6.404516</td>\n",
       "      <td>6.751566</td>\n",
       "      <td>9.289758</td>\n",
       "      <td>11.698169</td>\n",
       "      <td>10.306115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MB-0008</td>\n",
       "      <td>1</td>\n",
       "      <td>76.97</td>\n",
       "      <td>41.366667</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "      <td>40.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>2.0</td>\n",
       "      <td>11.276581</td>\n",
       "      <td>9.956267</td>\n",
       "      <td>7.331223</td>\n",
       "      <td>6.337951</td>\n",
       "      <td>5.817818</td>\n",
       "      <td>9.719781</td>\n",
       "      <td>11.625006</td>\n",
       "      <td>12.161961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MB-0010</td>\n",
       "      <td>1</td>\n",
       "      <td>78.77</td>\n",
       "      <td>7.800000</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>NO</td>\n",
       "      <td>YES</td>\n",
       "      <td>31.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>4.0</td>\n",
       "      <td>11.239750</td>\n",
       "      <td>9.739996</td>\n",
       "      <td>5.954311</td>\n",
       "      <td>5.419711</td>\n",
       "      <td>6.123056</td>\n",
       "      <td>9.787085</td>\n",
       "      <td>12.142178</td>\n",
       "      <td>11.433164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MB-0035</td>\n",
       "      <td>1</td>\n",
       "      <td>84.22</td>\n",
       "      <td>36.266667</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>28.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.536847</td>\n",
       "      <td>11.514514</td>\n",
       "      <td>5.587666</td>\n",
       "      <td>6.411477</td>\n",
       "      <td>5.988243</td>\n",
       "      <td>10.212610</td>\n",
       "      <td>12.804542</td>\n",
       "      <td>13.474571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MB-0036</td>\n",
       "      <td>1</td>\n",
       "      <td>85.49</td>\n",
       "      <td>132.033333</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>NO</td>\n",
       "      <td>YES</td>\n",
       "      <td>22.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.927313</td>\n",
       "      <td>10.033753</td>\n",
       "      <td>7.002502</td>\n",
       "      <td>5.943520</td>\n",
       "      <td>5.650386</td>\n",
       "      <td>10.845709</td>\n",
       "      <td>11.991655</td>\n",
       "      <td>12.466928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1895</th>\n",
       "      <td>MB-7291</td>\n",
       "      <td>4</td>\n",
       "      <td>53.87</td>\n",
       "      <td>6.833333</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>50.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.096158</td>\n",
       "      <td>12.600238</td>\n",
       "      <td>5.260572</td>\n",
       "      <td>6.351516</td>\n",
       "      <td>5.829438</td>\n",
       "      <td>8.116684</td>\n",
       "      <td>10.298983</td>\n",
       "      <td>10.747426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1896</th>\n",
       "      <td>MB-7292</td>\n",
       "      <td>4</td>\n",
       "      <td>52.90</td>\n",
       "      <td>78.466667</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>NO</td>\n",
       "      <td>YES</td>\n",
       "      <td>15.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>3.0</td>\n",
       "      <td>9.565225</td>\n",
       "      <td>10.215534</td>\n",
       "      <td>6.946029</td>\n",
       "      <td>6.230146</td>\n",
       "      <td>5.930059</td>\n",
       "      <td>8.844543</td>\n",
       "      <td>10.949128</td>\n",
       "      <td>10.711007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1898</th>\n",
       "      <td>MB-7294</td>\n",
       "      <td>4</td>\n",
       "      <td>59.20</td>\n",
       "      <td>82.733333</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>15.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>11.290976</td>\n",
       "      <td>10.846545</td>\n",
       "      <td>7.312247</td>\n",
       "      <td>5.660943</td>\n",
       "      <td>6.190000</td>\n",
       "      <td>9.424235</td>\n",
       "      <td>11.075688</td>\n",
       "      <td>11.567166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1900</th>\n",
       "      <td>MB-7296</td>\n",
       "      <td>4</td>\n",
       "      <td>42.88</td>\n",
       "      <td>44.733333</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>NO</td>\n",
       "      <td>YES</td>\n",
       "      <td>20.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9.733986</td>\n",
       "      <td>13.753037</td>\n",
       "      <td>5.616082</td>\n",
       "      <td>6.271912</td>\n",
       "      <td>5.999093</td>\n",
       "      <td>9.530390</td>\n",
       "      <td>11.532033</td>\n",
       "      <td>11.626140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1901</th>\n",
       "      <td>MB-7297</td>\n",
       "      <td>4</td>\n",
       "      <td>62.90</td>\n",
       "      <td>175.966667</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>NO</td>\n",
       "      <td>YES</td>\n",
       "      <td>25.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>11.053198</td>\n",
       "      <td>10.228570</td>\n",
       "      <td>7.478069</td>\n",
       "      <td>6.212256</td>\n",
       "      <td>6.192399</td>\n",
       "      <td>9.540589</td>\n",
       "      <td>11.482761</td>\n",
       "      <td>11.180360</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>622 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Patient_ID  Cohort  Age_at_diagnosis  Survival_time Survival_status  \\\n",
       "2       MB-0005       1             48.87     163.700000        DECEASED   \n",
       "4       MB-0008       1             76.97      41.366667        DECEASED   \n",
       "5       MB-0010       1             78.77       7.800000        DECEASED   \n",
       "9       MB-0035       1             84.22      36.266667        DECEASED   \n",
       "10      MB-0036       1             85.49     132.033333        DECEASED   \n",
       "...         ...     ...               ...            ...             ...   \n",
       "1895    MB-7291       4             53.87       6.833333        DECEASED   \n",
       "1896    MB-7292       4             52.90      78.466667        DECEASED   \n",
       "1898    MB-7294       4             59.20      82.733333        DECEASED   \n",
       "1900    MB-7296       4             42.88      44.733333        DECEASED   \n",
       "1901    MB-7297       4             62.90     175.966667        DECEASED   \n",
       "\n",
       "         Vital_status Chemotherapy Radiotherapy  Tumour_size  Tumour_stage  \\\n",
       "2     Died of Disease          YES           NO         15.0           2.0   \n",
       "4     Died of Disease          YES          YES         40.0           2.0   \n",
       "5     Died of Disease           NO          YES         31.0           4.0   \n",
       "9     Died of Disease           NO           NO         28.0           2.0   \n",
       "10    Died of Disease           NO          YES         22.0           4.0   \n",
       "...               ...          ...          ...          ...           ...   \n",
       "1895  Died of Disease           NO           NO         50.0           NaN   \n",
       "1896  Died of Disease           NO          YES         15.0           NaN   \n",
       "1898  Died of Disease           NO           NO         15.0           NaN   \n",
       "1900  Died of Disease           NO          YES         20.0           NaN   \n",
       "1901  Died of Disease           NO          YES         25.0           NaN   \n",
       "\n",
       "      ...  Integrative_cluster  Mutation_count       ESR1      ERBB2  \\\n",
       "2     ...                    3             2.0  10.041281   9.725825   \n",
       "4     ...                    9             2.0  11.276581   9.956267   \n",
       "5     ...                    7             4.0  11.239750   9.739996   \n",
       "9     ...                    3             5.0   7.536847  11.514514   \n",
       "10    ...                    3             1.0  10.927313  10.033753   \n",
       "...   ...                  ...             ...        ...        ...   \n",
       "1895  ...                    1             4.0   6.096158  12.600238   \n",
       "1896  ...                    6             3.0   9.565225  10.215534   \n",
       "1898  ...                    1             2.0  11.290976  10.846545   \n",
       "1900  ...                    5             6.0   9.733986  13.753037   \n",
       "1901  ...                    1             4.0  11.053198  10.228570   \n",
       "\n",
       "           PGR      TP53    PIK3CA      GATA3      FOXA1       MLPH  \n",
       "2     7.376123  6.404516  6.751566   9.289758  11.698169  10.306115  \n",
       "4     7.331223  6.337951  5.817818   9.719781  11.625006  12.161961  \n",
       "5     5.954311  5.419711  6.123056   9.787085  12.142178  11.433164  \n",
       "9     5.587666  6.411477  5.988243  10.212610  12.804542  13.474571  \n",
       "10    7.002502  5.943520  5.650386  10.845709  11.991655  12.466928  \n",
       "...        ...       ...       ...        ...        ...        ...  \n",
       "1895  5.260572  6.351516  5.829438   8.116684  10.298983  10.747426  \n",
       "1896  6.946029  6.230146  5.930059   8.844543  10.949128  10.711007  \n",
       "1898  7.312247  5.660943  6.190000   9.424235  11.075688  11.567166  \n",
       "1900  5.616082  6.271912  5.999093   9.530390  11.532033  11.626140  \n",
       "1901  7.478069  6.212256  6.192399   9.540589  11.482761  11.180360  \n",
       "\n",
       "[622 rows x 32 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabric[metabric.Vital_status==\"Died of Disease\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To filter based on more than one condition, you can use the operators `&` (and), `|` (or). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Cohort</th>\n",
       "      <th>Age_at_diagnosis</th>\n",
       "      <th>Survival_time</th>\n",
       "      <th>Survival_status</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>Chemotherapy</th>\n",
       "      <th>Radiotherapy</th>\n",
       "      <th>Tumour_size</th>\n",
       "      <th>Tumour_stage</th>\n",
       "      <th>...</th>\n",
       "      <th>Integrative_cluster</th>\n",
       "      <th>Mutation_count</th>\n",
       "      <th>ESR1</th>\n",
       "      <th>ERBB2</th>\n",
       "      <th>PGR</th>\n",
       "      <th>TP53</th>\n",
       "      <th>PIK3CA</th>\n",
       "      <th>GATA3</th>\n",
       "      <th>FOXA1</th>\n",
       "      <th>MLPH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MB-0008</td>\n",
       "      <td>1</td>\n",
       "      <td>76.97</td>\n",
       "      <td>41.366667</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "      <td>40.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>2.0</td>\n",
       "      <td>11.276581</td>\n",
       "      <td>9.956267</td>\n",
       "      <td>7.331223</td>\n",
       "      <td>6.337951</td>\n",
       "      <td>5.817818</td>\n",
       "      <td>9.719781</td>\n",
       "      <td>11.625006</td>\n",
       "      <td>12.161961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MB-0010</td>\n",
       "      <td>1</td>\n",
       "      <td>78.77</td>\n",
       "      <td>7.800000</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>NO</td>\n",
       "      <td>YES</td>\n",
       "      <td>31.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>4.0</td>\n",
       "      <td>11.239750</td>\n",
       "      <td>9.739996</td>\n",
       "      <td>5.954311</td>\n",
       "      <td>5.419711</td>\n",
       "      <td>6.123056</td>\n",
       "      <td>9.787085</td>\n",
       "      <td>12.142178</td>\n",
       "      <td>11.433164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MB-0035</td>\n",
       "      <td>1</td>\n",
       "      <td>84.22</td>\n",
       "      <td>36.266667</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>28.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.536847</td>\n",
       "      <td>11.514514</td>\n",
       "      <td>5.587666</td>\n",
       "      <td>6.411477</td>\n",
       "      <td>5.988243</td>\n",
       "      <td>10.212610</td>\n",
       "      <td>12.804542</td>\n",
       "      <td>13.474571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MB-0036</td>\n",
       "      <td>1</td>\n",
       "      <td>85.49</td>\n",
       "      <td>132.033333</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>NO</td>\n",
       "      <td>YES</td>\n",
       "      <td>22.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.927313</td>\n",
       "      <td>10.033753</td>\n",
       "      <td>7.002502</td>\n",
       "      <td>5.943520</td>\n",
       "      <td>5.650386</td>\n",
       "      <td>10.845709</td>\n",
       "      <td>11.991655</td>\n",
       "      <td>12.466928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>MB-0112</td>\n",
       "      <td>1</td>\n",
       "      <td>83.89</td>\n",
       "      <td>39.166667</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>NO</td>\n",
       "      <td>YES</td>\n",
       "      <td>150.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.185200</td>\n",
       "      <td>9.580607</td>\n",
       "      <td>5.480888</td>\n",
       "      <td>6.068563</td>\n",
       "      <td>5.513442</td>\n",
       "      <td>10.334501</td>\n",
       "      <td>11.513395</td>\n",
       "      <td>12.185778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1786</th>\n",
       "      <td>MB-7133</td>\n",
       "      <td>4</td>\n",
       "      <td>71.41</td>\n",
       "      <td>23.833333</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>NO</td>\n",
       "      <td>YES</td>\n",
       "      <td>18.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.963847</td>\n",
       "      <td>10.486243</td>\n",
       "      <td>5.446796</td>\n",
       "      <td>5.592907</td>\n",
       "      <td>5.910852</td>\n",
       "      <td>10.360905</td>\n",
       "      <td>11.304980</td>\n",
       "      <td>11.840719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1823</th>\n",
       "      <td>MB-7185</td>\n",
       "      <td>4</td>\n",
       "      <td>70.76</td>\n",
       "      <td>125.033333</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>NO</td>\n",
       "      <td>YES</td>\n",
       "      <td>16.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11.949865</td>\n",
       "      <td>11.485468</td>\n",
       "      <td>6.465373</td>\n",
       "      <td>6.495243</td>\n",
       "      <td>5.635878</td>\n",
       "      <td>10.844879</td>\n",
       "      <td>12.333075</td>\n",
       "      <td>12.708275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1858</th>\n",
       "      <td>MB-7237</td>\n",
       "      <td>4</td>\n",
       "      <td>77.37</td>\n",
       "      <td>23.933333</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>18.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>10.0</td>\n",
       "      <td>9.879010</td>\n",
       "      <td>11.007182</td>\n",
       "      <td>6.259103</td>\n",
       "      <td>6.518297</td>\n",
       "      <td>5.678203</td>\n",
       "      <td>10.594228</td>\n",
       "      <td>10.749515</td>\n",
       "      <td>12.119799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1859</th>\n",
       "      <td>MB-7238</td>\n",
       "      <td>4</td>\n",
       "      <td>71.11</td>\n",
       "      <td>23.266667</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>27.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>7.0</td>\n",
       "      <td>9.520532</td>\n",
       "      <td>10.370414</td>\n",
       "      <td>5.390623</td>\n",
       "      <td>5.771310</td>\n",
       "      <td>5.560965</td>\n",
       "      <td>10.110271</td>\n",
       "      <td>11.426720</td>\n",
       "      <td>12.080146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1885</th>\n",
       "      <td>MB-7279</td>\n",
       "      <td>4</td>\n",
       "      <td>86.14</td>\n",
       "      <td>21.300000</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>50.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.377438</td>\n",
       "      <td>12.237637</td>\n",
       "      <td>6.495096</td>\n",
       "      <td>6.568044</td>\n",
       "      <td>5.314676</td>\n",
       "      <td>9.369560</td>\n",
       "      <td>11.456438</td>\n",
       "      <td>12.027578</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>158 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Patient_ID  Cohort  Age_at_diagnosis  Survival_time Survival_status  \\\n",
       "4       MB-0008       1             76.97      41.366667        DECEASED   \n",
       "5       MB-0010       1             78.77       7.800000        DECEASED   \n",
       "9       MB-0035       1             84.22      36.266667        DECEASED   \n",
       "10      MB-0036       1             85.49     132.033333        DECEASED   \n",
       "38      MB-0112       1             83.89      39.166667        DECEASED   \n",
       "...         ...     ...               ...            ...             ...   \n",
       "1786    MB-7133       4             71.41      23.833333        DECEASED   \n",
       "1823    MB-7185       4             70.76     125.033333        DECEASED   \n",
       "1858    MB-7237       4             77.37      23.933333        DECEASED   \n",
       "1859    MB-7238       4             71.11      23.266667        DECEASED   \n",
       "1885    MB-7279       4             86.14      21.300000        DECEASED   \n",
       "\n",
       "         Vital_status Chemotherapy Radiotherapy  Tumour_size  Tumour_stage  \\\n",
       "4     Died of Disease          YES          YES         40.0           2.0   \n",
       "5     Died of Disease           NO          YES         31.0           4.0   \n",
       "9     Died of Disease           NO           NO         28.0           2.0   \n",
       "10    Died of Disease           NO          YES         22.0           4.0   \n",
       "38    Died of Disease           NO          YES        150.0           3.0   \n",
       "...               ...          ...          ...          ...           ...   \n",
       "1786  Died of Disease           NO          YES         18.0           NaN   \n",
       "1823  Died of Disease           NO          YES         16.2           NaN   \n",
       "1858  Died of Disease           NO           NO         18.0           NaN   \n",
       "1859  Died of Disease           NO           NO         27.0           NaN   \n",
       "1885  Died of Disease           NO           NO         50.0           NaN   \n",
       "\n",
       "      ...  Integrative_cluster  Mutation_count       ESR1      ERBB2  \\\n",
       "4     ...                    9             2.0  11.276581   9.956267   \n",
       "5     ...                    7             4.0  11.239750   9.739996   \n",
       "9     ...                    3             5.0   7.536847  11.514514   \n",
       "10    ...                    3             1.0  10.927313  10.033753   \n",
       "38    ...                    3             8.0   9.185200   9.580607   \n",
       "...   ...                  ...             ...        ...        ...   \n",
       "1786  ...                    1             8.0   9.963847  10.486243   \n",
       "1823  ...                    8             9.0  11.949865  11.485468   \n",
       "1858  ...                    9            10.0   9.879010  11.007182   \n",
       "1859  ...                    8             7.0   9.520532  10.370414   \n",
       "1885  ...                    9             4.0   9.377438  12.237637   \n",
       "\n",
       "           PGR      TP53    PIK3CA      GATA3      FOXA1       MLPH  \n",
       "4     7.331223  6.337951  5.817818   9.719781  11.625006  12.161961  \n",
       "5     5.954311  5.419711  6.123056   9.787085  12.142178  11.433164  \n",
       "9     5.587666  6.411477  5.988243  10.212610  12.804542  13.474571  \n",
       "10    7.002502  5.943520  5.650386  10.845709  11.991655  12.466928  \n",
       "38    5.480888  6.068563  5.513442  10.334501  11.513395  12.185778  \n",
       "...        ...       ...       ...        ...        ...        ...  \n",
       "1786  5.446796  5.592907  5.910852  10.360905  11.304980  11.840719  \n",
       "1823  6.465373  6.495243  5.635878  10.844879  12.333075  12.708275  \n",
       "1858  6.259103  6.518297  5.678203  10.594228  10.749515  12.119799  \n",
       "1859  5.390623  5.771310  5.560965  10.110271  11.426720  12.080146  \n",
       "1885  6.495096  6.568044  5.314676   9.369560  11.456438  12.027578  \n",
       "\n",
       "[158 rows x 32 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabric[(metabric.Vital_status==\"Died of Disease\") & (metabric.Age_at_diagnosis>70)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For categorical variables e.g. `Vital_status` or `Cohort`, it may be useful to count how many occurrences there is for each category:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metabric['Vital_status'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metabric['Vital_status'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To filter by more than one category, use the `.isin()` method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metabric[metabric.Vital_status.isin(['Died of Disease', 'Died of Other Causes'])]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sort data\n",
    "\n",
    "To sort the entire DataFrame according to one of the columns, we can use the `.sort_values()` method. We can store the sorted DataFrame using a new variable name such as `metabric_sorted`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Cohort</th>\n",
       "      <th>Age_at_diagnosis</th>\n",
       "      <th>Survival_time</th>\n",
       "      <th>Survival_status</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>Chemotherapy</th>\n",
       "      <th>Radiotherapy</th>\n",
       "      <th>Tumour_size</th>\n",
       "      <th>Tumour_stage</th>\n",
       "      <th>...</th>\n",
       "      <th>Integrative_cluster</th>\n",
       "      <th>Mutation_count</th>\n",
       "      <th>ESR1</th>\n",
       "      <th>ERBB2</th>\n",
       "      <th>PGR</th>\n",
       "      <th>TP53</th>\n",
       "      <th>PIK3CA</th>\n",
       "      <th>GATA3</th>\n",
       "      <th>FOXA1</th>\n",
       "      <th>MLPH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1178</th>\n",
       "      <td>MB-5154</td>\n",
       "      <td>3</td>\n",
       "      <td>31.87</td>\n",
       "      <td>170.833333</td>\n",
       "      <td>LIVING</td>\n",
       "      <td>Living</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.390271</td>\n",
       "      <td>14.150505</td>\n",
       "      <td>5.374285</td>\n",
       "      <td>6.166627</td>\n",
       "      <td>5.730209</td>\n",
       "      <td>7.682715</td>\n",
       "      <td>10.980236</td>\n",
       "      <td>11.920582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>939</th>\n",
       "      <td>MB-4714</td>\n",
       "      <td>3</td>\n",
       "      <td>64.68</td>\n",
       "      <td>83.666667</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Other Causes</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5.756730</td>\n",
       "      <td>11.451638</td>\n",
       "      <td>5.218476</td>\n",
       "      <td>6.951965</td>\n",
       "      <td>6.113398</td>\n",
       "      <td>7.928369</td>\n",
       "      <td>5.675331</td>\n",
       "      <td>7.096104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1489</th>\n",
       "      <td>MB-5640</td>\n",
       "      <td>3</td>\n",
       "      <td>58.00</td>\n",
       "      <td>172.300000</td>\n",
       "      <td>LIVING</td>\n",
       "      <td>Living</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4ER+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>11.471468</td>\n",
       "      <td>10.710609</td>\n",
       "      <td>6.767783</td>\n",
       "      <td>5.857385</td>\n",
       "      <td>5.779344</td>\n",
       "      <td>10.733006</td>\n",
       "      <td>12.006139</td>\n",
       "      <td>12.380562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>867</th>\n",
       "      <td>MB-4548</td>\n",
       "      <td>3</td>\n",
       "      <td>50.42</td>\n",
       "      <td>322.833333</td>\n",
       "      <td>LIVING</td>\n",
       "      <td>Living</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4ER-</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.703593</td>\n",
       "      <td>13.999965</td>\n",
       "      <td>5.270841</td>\n",
       "      <td>6.500610</td>\n",
       "      <td>6.011818</td>\n",
       "      <td>7.070812</td>\n",
       "      <td>8.922047</td>\n",
       "      <td>10.536358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1343</th>\n",
       "      <td>MB-5407</td>\n",
       "      <td>3</td>\n",
       "      <td>36.03</td>\n",
       "      <td>205.600000</td>\n",
       "      <td>LIVING</td>\n",
       "      <td>Living</td>\n",
       "      <td>NO</td>\n",
       "      <td>YES</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>9.0</td>\n",
       "      <td>10.877464</td>\n",
       "      <td>11.983508</td>\n",
       "      <td>7.275882</td>\n",
       "      <td>6.079454</td>\n",
       "      <td>5.890194</td>\n",
       "      <td>11.062042</td>\n",
       "      <td>11.983090</td>\n",
       "      <td>12.856245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1568</th>\n",
       "      <td>MB-6141</td>\n",
       "      <td>5</td>\n",
       "      <td>79.84</td>\n",
       "      <td>78.866667</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Other Causes</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.155460</td>\n",
       "      <td>10.711909</td>\n",
       "      <td>5.362010</td>\n",
       "      <td>5.913182</td>\n",
       "      <td>5.827326</td>\n",
       "      <td>11.645141</td>\n",
       "      <td>11.964034</td>\n",
       "      <td>13.320232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1580</th>\n",
       "      <td>MB-6160</td>\n",
       "      <td>5</td>\n",
       "      <td>46.23</td>\n",
       "      <td>247.366667</td>\n",
       "      <td>LIVING</td>\n",
       "      <td>Living</td>\n",
       "      <td>YES</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.887579</td>\n",
       "      <td>13.896580</td>\n",
       "      <td>5.494239</td>\n",
       "      <td>6.122002</td>\n",
       "      <td>5.795817</td>\n",
       "      <td>8.259371</td>\n",
       "      <td>11.106729</td>\n",
       "      <td>10.908745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1609</th>\n",
       "      <td>MB-6217</td>\n",
       "      <td>5</td>\n",
       "      <td>47.51</td>\n",
       "      <td>211.533333</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Other Causes</td>\n",
       "      <td>YES</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>6.0</td>\n",
       "      <td>10.312296</td>\n",
       "      <td>11.494632</td>\n",
       "      <td>6.158972</td>\n",
       "      <td>6.177842</td>\n",
       "      <td>5.655659</td>\n",
       "      <td>10.679476</td>\n",
       "      <td>11.568274</td>\n",
       "      <td>12.442291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1620</th>\n",
       "      <td>MB-6234</td>\n",
       "      <td>5</td>\n",
       "      <td>66.35</td>\n",
       "      <td>213.366667</td>\n",
       "      <td>LIVING</td>\n",
       "      <td>Living</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>4ER+</td>\n",
       "      <td>4.0</td>\n",
       "      <td>10.801931</td>\n",
       "      <td>10.695930</td>\n",
       "      <td>6.603785</td>\n",
       "      <td>5.957729</td>\n",
       "      <td>5.790894</td>\n",
       "      <td>10.569577</td>\n",
       "      <td>11.251508</td>\n",
       "      <td>11.758685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1635</th>\n",
       "      <td>MB-6272</td>\n",
       "      <td>5</td>\n",
       "      <td>36.93</td>\n",
       "      <td>22.400000</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.741627</td>\n",
       "      <td>10.298811</td>\n",
       "      <td>5.291367</td>\n",
       "      <td>5.701960</td>\n",
       "      <td>5.747574</td>\n",
       "      <td>6.121951</td>\n",
       "      <td>5.984301</td>\n",
       "      <td>6.368798</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1904 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Patient_ID  Cohort  Age_at_diagnosis  Survival_time Survival_status  \\\n",
       "1178    MB-5154       3             31.87     170.833333          LIVING   \n",
       "939     MB-4714       3             64.68      83.666667        DECEASED   \n",
       "1489    MB-5640       3             58.00     172.300000          LIVING   \n",
       "867     MB-4548       3             50.42     322.833333          LIVING   \n",
       "1343    MB-5407       3             36.03     205.600000          LIVING   \n",
       "...         ...     ...               ...            ...             ...   \n",
       "1568    MB-6141       5             79.84      78.866667        DECEASED   \n",
       "1580    MB-6160       5             46.23     247.366667          LIVING   \n",
       "1609    MB-6217       5             47.51     211.533333        DECEASED   \n",
       "1620    MB-6234       5             66.35     213.366667          LIVING   \n",
       "1635    MB-6272       5             36.93      22.400000        DECEASED   \n",
       "\n",
       "              Vital_status Chemotherapy Radiotherapy  Tumour_size  \\\n",
       "1178                Living           NO           NO          1.0   \n",
       "939   Died of Other Causes          YES          YES          1.0   \n",
       "1489                Living           NO           NO          1.0   \n",
       "867                 Living          YES          YES          1.0   \n",
       "1343                Living           NO          YES          1.0   \n",
       "...                    ...          ...          ...          ...   \n",
       "1568  Died of Other Causes           NO           NO          NaN   \n",
       "1580                Living          YES           NO          NaN   \n",
       "1609  Died of Other Causes          YES           NO          NaN   \n",
       "1620                Living           NO           NO          NaN   \n",
       "1635       Died of Disease          YES          YES          NaN   \n",
       "\n",
       "      Tumour_stage  ...  Integrative_cluster  Mutation_count       ESR1  \\\n",
       "1178           1.0  ...                    3             2.0   6.390271   \n",
       "939            2.0  ...                   10             7.0   5.756730   \n",
       "1489           1.0  ...                 4ER+             5.0  11.471468   \n",
       "867            2.0  ...                 4ER-             2.0   5.703593   \n",
       "1343           1.0  ...                    8             9.0  10.877464   \n",
       "...            ...  ...                  ...             ...        ...   \n",
       "1568           NaN  ...                    8             1.0  11.155460   \n",
       "1580           NaN  ...                    9             4.0   5.887579   \n",
       "1609           NaN  ...                    9             6.0  10.312296   \n",
       "1620           NaN  ...                 4ER+             4.0  10.801931   \n",
       "1635           NaN  ...                   10             6.0   5.741627   \n",
       "\n",
       "          ERBB2       PGR      TP53    PIK3CA      GATA3      FOXA1       MLPH  \n",
       "1178  14.150505  5.374285  6.166627  5.730209   7.682715  10.980236  11.920582  \n",
       "939   11.451638  5.218476  6.951965  6.113398   7.928369   5.675331   7.096104  \n",
       "1489  10.710609  6.767783  5.857385  5.779344  10.733006  12.006139  12.380562  \n",
       "867   13.999965  5.270841  6.500610  6.011818   7.070812   8.922047  10.536358  \n",
       "1343  11.983508  7.275882  6.079454  5.890194  11.062042  11.983090  12.856245  \n",
       "...         ...       ...       ...       ...        ...        ...        ...  \n",
       "1568  10.711909  5.362010  5.913182  5.827326  11.645141  11.964034  13.320232  \n",
       "1580  13.896580  5.494239  6.122002  5.795817   8.259371  11.106729  10.908745  \n",
       "1609  11.494632  6.158972  6.177842  5.655659  10.679476  11.568274  12.442291  \n",
       "1620  10.695930  6.603785  5.957729  5.790894  10.569577  11.251508  11.758685  \n",
       "1635  10.298811  5.291367  5.701960  5.747574   6.121951   5.984301   6.368798  \n",
       "\n",
       "[1904 rows x 32 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabric_sorted = metabric.sort_values('Tumour_size')\n",
    "metabric_sorted"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also sort the DataFrame in descending order:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Cohort</th>\n",
       "      <th>Age_at_diagnosis</th>\n",
       "      <th>Survival_time</th>\n",
       "      <th>Survival_status</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>Chemotherapy</th>\n",
       "      <th>Radiotherapy</th>\n",
       "      <th>Tumour_size</th>\n",
       "      <th>Tumour_stage</th>\n",
       "      <th>...</th>\n",
       "      <th>Integrative_cluster</th>\n",
       "      <th>Mutation_count</th>\n",
       "      <th>ESR1</th>\n",
       "      <th>ERBB2</th>\n",
       "      <th>PGR</th>\n",
       "      <th>TP53</th>\n",
       "      <th>PIK3CA</th>\n",
       "      <th>GATA3</th>\n",
       "      <th>FOXA1</th>\n",
       "      <th>MLPH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1812</th>\n",
       "      <td>MB-7165</td>\n",
       "      <td>4</td>\n",
       "      <td>28.78</td>\n",
       "      <td>147.166667</td>\n",
       "      <td>LIVING</td>\n",
       "      <td>Living</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "      <td>182.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5.426713</td>\n",
       "      <td>9.507019</td>\n",
       "      <td>5.229264</td>\n",
       "      <td>7.921411</td>\n",
       "      <td>6.407379</td>\n",
       "      <td>7.674084</td>\n",
       "      <td>7.179491</td>\n",
       "      <td>9.213541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>MB-0406</td>\n",
       "      <td>1</td>\n",
       "      <td>69.32</td>\n",
       "      <td>96.900000</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "      <td>180.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>5.0</td>\n",
       "      <td>10.295546</td>\n",
       "      <td>10.515396</td>\n",
       "      <td>5.867643</td>\n",
       "      <td>6.085246</td>\n",
       "      <td>5.780428</td>\n",
       "      <td>10.889048</td>\n",
       "      <td>11.546029</td>\n",
       "      <td>12.883059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460</th>\n",
       "      <td>MB-0660</td>\n",
       "      <td>1</td>\n",
       "      <td>52.43</td>\n",
       "      <td>18.933333</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "      <td>160.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.806700</td>\n",
       "      <td>10.392012</td>\n",
       "      <td>5.534153</td>\n",
       "      <td>6.319585</td>\n",
       "      <td>5.647157</td>\n",
       "      <td>9.137186</td>\n",
       "      <td>11.317307</td>\n",
       "      <td>10.730205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>MB-0112</td>\n",
       "      <td>1</td>\n",
       "      <td>83.89</td>\n",
       "      <td>39.166667</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>NO</td>\n",
       "      <td>YES</td>\n",
       "      <td>150.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.185200</td>\n",
       "      <td>9.580607</td>\n",
       "      <td>5.480888</td>\n",
       "      <td>6.068563</td>\n",
       "      <td>5.513442</td>\n",
       "      <td>10.334501</td>\n",
       "      <td>11.513395</td>\n",
       "      <td>12.185778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>710</th>\n",
       "      <td>MB-3436</td>\n",
       "      <td>2</td>\n",
       "      <td>38.20</td>\n",
       "      <td>50.766667</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "      <td>130.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>10.227027</td>\n",
       "      <td>9.888552</td>\n",
       "      <td>8.437960</td>\n",
       "      <td>6.322436</td>\n",
       "      <td>6.534325</td>\n",
       "      <td>9.243268</td>\n",
       "      <td>12.050932</td>\n",
       "      <td>12.602092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1568</th>\n",
       "      <td>MB-6141</td>\n",
       "      <td>5</td>\n",
       "      <td>79.84</td>\n",
       "      <td>78.866667</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Other Causes</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.155460</td>\n",
       "      <td>10.711909</td>\n",
       "      <td>5.362010</td>\n",
       "      <td>5.913182</td>\n",
       "      <td>5.827326</td>\n",
       "      <td>11.645141</td>\n",
       "      <td>11.964034</td>\n",
       "      <td>13.320232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1580</th>\n",
       "      <td>MB-6160</td>\n",
       "      <td>5</td>\n",
       "      <td>46.23</td>\n",
       "      <td>247.366667</td>\n",
       "      <td>LIVING</td>\n",
       "      <td>Living</td>\n",
       "      <td>YES</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.887579</td>\n",
       "      <td>13.896580</td>\n",
       "      <td>5.494239</td>\n",
       "      <td>6.122002</td>\n",
       "      <td>5.795817</td>\n",
       "      <td>8.259371</td>\n",
       "      <td>11.106729</td>\n",
       "      <td>10.908745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1609</th>\n",
       "      <td>MB-6217</td>\n",
       "      <td>5</td>\n",
       "      <td>47.51</td>\n",
       "      <td>211.533333</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Other Causes</td>\n",
       "      <td>YES</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>6.0</td>\n",
       "      <td>10.312296</td>\n",
       "      <td>11.494632</td>\n",
       "      <td>6.158972</td>\n",
       "      <td>6.177842</td>\n",
       "      <td>5.655659</td>\n",
       "      <td>10.679476</td>\n",
       "      <td>11.568274</td>\n",
       "      <td>12.442291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1620</th>\n",
       "      <td>MB-6234</td>\n",
       "      <td>5</td>\n",
       "      <td>66.35</td>\n",
       "      <td>213.366667</td>\n",
       "      <td>LIVING</td>\n",
       "      <td>Living</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>4ER+</td>\n",
       "      <td>4.0</td>\n",
       "      <td>10.801931</td>\n",
       "      <td>10.695930</td>\n",
       "      <td>6.603785</td>\n",
       "      <td>5.957729</td>\n",
       "      <td>5.790894</td>\n",
       "      <td>10.569577</td>\n",
       "      <td>11.251508</td>\n",
       "      <td>11.758685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1635</th>\n",
       "      <td>MB-6272</td>\n",
       "      <td>5</td>\n",
       "      <td>36.93</td>\n",
       "      <td>22.400000</td>\n",
       "      <td>DECEASED</td>\n",
       "      <td>Died of Disease</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.741627</td>\n",
       "      <td>10.298811</td>\n",
       "      <td>5.291367</td>\n",
       "      <td>5.701960</td>\n",
       "      <td>5.747574</td>\n",
       "      <td>6.121951</td>\n",
       "      <td>5.984301</td>\n",
       "      <td>6.368798</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1904 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Patient_ID  Cohort  Age_at_diagnosis  Survival_time Survival_status  \\\n",
       "1812    MB-7165       4             28.78     147.166667          LIVING   \n",
       "266     MB-0406       1             69.32      96.900000        DECEASED   \n",
       "460     MB-0660       1             52.43      18.933333        DECEASED   \n",
       "38      MB-0112       1             83.89      39.166667        DECEASED   \n",
       "710     MB-3436       2             38.20      50.766667        DECEASED   \n",
       "...         ...     ...               ...            ...             ...   \n",
       "1568    MB-6141       5             79.84      78.866667        DECEASED   \n",
       "1580    MB-6160       5             46.23     247.366667          LIVING   \n",
       "1609    MB-6217       5             47.51     211.533333        DECEASED   \n",
       "1620    MB-6234       5             66.35     213.366667          LIVING   \n",
       "1635    MB-6272       5             36.93      22.400000        DECEASED   \n",
       "\n",
       "              Vital_status Chemotherapy Radiotherapy  Tumour_size  \\\n",
       "1812                Living          YES          YES        182.0   \n",
       "266        Died of Disease          YES          YES        180.0   \n",
       "460        Died of Disease          YES          YES        160.0   \n",
       "38         Died of Disease           NO          YES        150.0   \n",
       "710        Died of Disease          YES          YES        130.0   \n",
       "...                    ...          ...          ...          ...   \n",
       "1568  Died of Other Causes           NO           NO          NaN   \n",
       "1580                Living          YES           NO          NaN   \n",
       "1609  Died of Other Causes          YES           NO          NaN   \n",
       "1620                Living           NO           NO          NaN   \n",
       "1635       Died of Disease          YES          YES          NaN   \n",
       "\n",
       "      Tumour_stage  ...  Integrative_cluster  Mutation_count       ESR1  \\\n",
       "1812           NaN  ...                   10             7.0   5.426713   \n",
       "266            3.0  ...                    2             5.0  10.295546   \n",
       "460            1.0  ...                   10             5.0   6.806700   \n",
       "38             3.0  ...                    3             8.0   9.185200   \n",
       "710            3.0  ...                    1             4.0  10.227027   \n",
       "...            ...  ...                  ...             ...        ...   \n",
       "1568           NaN  ...                    8             1.0  11.155460   \n",
       "1580           NaN  ...                    9             4.0   5.887579   \n",
       "1609           NaN  ...                    9             6.0  10.312296   \n",
       "1620           NaN  ...                 4ER+             4.0  10.801931   \n",
       "1635           NaN  ...                   10             6.0   5.741627   \n",
       "\n",
       "          ERBB2       PGR      TP53    PIK3CA      GATA3      FOXA1       MLPH  \n",
       "1812   9.507019  5.229264  7.921411  6.407379   7.674084   7.179491   9.213541  \n",
       "266   10.515396  5.867643  6.085246  5.780428  10.889048  11.546029  12.883059  \n",
       "460   10.392012  5.534153  6.319585  5.647157   9.137186  11.317307  10.730205  \n",
       "38     9.580607  5.480888  6.068563  5.513442  10.334501  11.513395  12.185778  \n",
       "710    9.888552  8.437960  6.322436  6.534325   9.243268  12.050932  12.602092  \n",
       "...         ...       ...       ...       ...        ...        ...        ...  \n",
       "1568  10.711909  5.362010  5.913182  5.827326  11.645141  11.964034  13.320232  \n",
       "1580  13.896580  5.494239  6.122002  5.795817   8.259371  11.106729  10.908745  \n",
       "1609  11.494632  6.158972  6.177842  5.655659  10.679476  11.568274  12.442291  \n",
       "1620  10.695930  6.603785  5.957729  5.790894  10.569577  11.251508  11.758685  \n",
       "1635  10.298811  5.291367  5.701960  5.747574   6.121951   5.984301   6.368798  \n",
       "\n",
       "[1904 rows x 32 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabric_sorted = metabric.sort_values('Tumour_size', ascending=False)\n",
    "metabric_sorted"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Missing data\n",
    "\n",
    "Pandas primarily uses `NaN` to represent missing data, which are by default not included in computations.\n",
    "\n",
    "The `.info()` method shown above already gave us a way to find columns containing missing data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metabric.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To get the locations where values are missing:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.isna(metabric)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metabric.isnull()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To drop any rows containing at least one column with missing data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metabric.dropna()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "However, from the other way around, to rather remove columns with at least one row with missing data, you need to use the 'axis' argument:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metabric.dropna(axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define in which columns to look for missing values before dropping the row:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metabric.dropna(subset = [\"Tumour_size\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metabric.dropna(subset = [\"Tumour_size\", \"Tumour_stage\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filling missing data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metabric.fillna(value=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metabric.fillna(value={'Tumour_size':0, 'Tumour_stage':5})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Grouping\n",
    "\n",
    "Grouping patients by Cohort and then applying the `.mean()` function to the resulting groups:\n",
    "\n",
    "*Note: There are more *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metabric.groupby('Cohort')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metabric.groupby('Cohort').mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Grouping by multiple columns forms a hierarchical index, and again we can apply the `.mean()` function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metabric.groupby(['Cohort', 'Vital_status']).mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Q&A\n",
    "\n",
    "### Duration: ~10'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "03b5de4342d12f63398c17f4fe4cb032ff67035882e35a656212d40f64756279"
  },
  "kernelspec": {
   "display_name": "Python 3.9.13 ('.venv': venv)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
